engine-types/src/types.rs

```diff
@@ -1,14 +1,129 @@
-use super::{str, vec, Add, Address, String, Sub, Vec, U256};
+use crate::{str, vec, Address, AsMut, AsRef, Deref, DerefMut, Display, String, Vec, U256};
 use borsh::{BorshDeserialize, BorshSerialize};
 
+use crate::fmt::Formatter;
+pub use evm_core::Opcode;
+
 pub type Balance = u128;
 pub type RawAddress = [u8; 20];
 pub type RawU256 = [u8; 32]; // Big-endian large integer type.
 pub type RawH256 = [u8; 32]; // Unformatted binary data of fixed length.
 pub type EthAddress = [u8; 20];
-pub type Gas = u64;
 pub type StorageUsage = u64;
 
+#[derive(Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+/// Near gas type which wraps an underlying u64.
+pub struct NearGas(u64);
+
+impl Display for NearGas {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Deref for NearGas {
+    type Target = u64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for NearGas {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl AsRef<u64> for NearGas {
+    fn as_ref(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl AsMut<u64> for NearGas {
+    fn as_mut(&mut self) -> &mut u64 {
+        &mut self.0
+    }
+}
+
+impl From<u64> for NearGas {
+    fn from(gas: u64) -> Self {
+        Self(gas)
+    }
+}
+
+impl NearGas {
+    /// Constructs a new `NearGas` with a given u64 value.
+    pub const fn new(gas: u64) -> NearGas {
+        Self(gas)
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self)
+    }
+
+    pub fn checked_mul(self, rhs: Self) -> Option<Self> {
```

Comment:
I do not think it semantically makes sense to be able to multiply gas values together. The whole point of creating a new type is to add semantics to `u64`; its not a pure number any more. When you multiple two things with units together you do not get the same kind of thing back again (for example multiplying two lengths gives an area, not a length).

I think specifically **not** including a multiplication impl is a good thing because there should never be any logic which requires that operation, so if you think you need it then either the logic is wrong or one quantity has been given the incorrect type.